subsys/bluetooth/mesh/adv.c

```diff
@@ -84,16 +89,103 @@ struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
 	(void)memset(adv, 0, sizeof(*adv));
 
 	adv->type         = type;
+	adv->tag          = tag;
 	adv->xmit         = xmit;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type, uint8_t xmit,
-				   k_timeout_t timeout)
+struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
+				   enum bt_mesh_adv_tag tag,
+				   uint8_t xmit, k_timeout_t timeout)
 {
 	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_alloc, type,
-					    xmit, timeout);
+					    tag, xmit, timeout);
+}
+
+#if CONFIG_BT_MESH_RELAY_ADV_SETS
+static struct net_buf *process_events(struct k_poll_event *ev, int count)
+{
+	for (; count; ev++, count--) {
+		BT_DBG("ev->state %u", ev->state);
+
+		switch (ev->state) {
+		case K_POLL_STATE_FIFO_DATA_AVAILABLE:
+			return net_buf_get(ev->fifo, K_NO_WAIT);
+		case K_POLL_STATE_NOT_READY:
+		case K_POLL_STATE_CANCELLED:
+			break;
+		default:
+			BT_WARN("Unexpected k_poll event state %u", ev->state);
+			break;
+		}
+	}
+
+	return NULL;
+}
+
+struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+{
+	int err;
+	struct k_poll_event events[] = {
+		K_POLL_EVENT_STATIC_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+						K_POLL_MODE_NOTIFY_ONLY,
+						&bt_mesh_adv_queue,
+						0),
+		K_POLL_EVENT_STATIC_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+						K_POLL_MODE_NOTIFY_ONLY,
+						&bt_mesh_relay_queue,
+						0),
+
+	};
+
+	err = k_poll(events, ARRAY_SIZE(events), timeout);
+	if (err) {
+		return NULL;
+	}
+
+	return process_events(events, ARRAY_SIZE(events));
+}
+
+static struct net_buf *bt_mesh_adv_buf_relay_get(k_timeout_t timeout)
+{
+	return net_buf_get(&bt_mesh_relay_queue, timeout);
+}
+
+struct net_buf *bt_mesh_adv_buf_get_by_tag(uint8_t tag, k_timeout_t timeout)
+{
+	if (!(tag ^ BT_MESH_ADV_TAG_RELAY)) {
```

Comment:
I really had to wrap my head around this too long, and I'm still not sure if I got the intention right. The more common way of checking if a specific bit (and nothing else) is set is `if ((val & MY_BIT) == MY_BIT)`. That said, I assume the purpose you're exclusively checking for that bit is that if anything else is set then that has higher priority. If that's the case, why don't you just order the if-statements so that the relay case becomes an `else if` statement?

Reply:
Great.